src/xref/odoc_xref.ml

```diff
@@ -21,22 +22,49 @@ type lookup_result = Component_table.lookup_unit_result =
   | Found of lookup_result_found
   | Not_found
 
+type msg = [ `Msg of string ]
+
+(** Internal. Used to handled errors from [fetch_unit] and [fetch_page] *)
+exception Fetch_failed of msg
+
 let core_types = Odoc_model.Predefined.core_types
 
 let core_exceptions = Odoc_model.Predefined.core_exceptions
 
 type resolver = Resolve.resolver
 
-let build_resolver = Resolve.build_resolver
+let build_resolver ?equal ?hash lookup_unit fetch_unit lookup_page fetch_page =
+  let fetch_unit root =
+    match fetch_unit root with
+    | Ok unit -> unit
+    | Error e -> raise (Fetch_failed e)
+  and fetch_page root =
+    match fetch_page root with
+    | Ok page -> page
+    | Error e -> raise (Fetch_failed e)
+  in
+  Resolve.build_resolver ?equal ?hash lookup_unit fetch_unit lookup_page fetch_page
 
-let resolve = Resolve.resolve
+let resolve resolver unit =
+  try Ok (Resolve.resolve resolver unit)
+  with Fetch_failed (`Msg _ as e) -> Error e
```

Comment:
I use an exception to propagate errors inside the resolver. This is to avoid a giant refactor and potentially making it slower.